onnxscript/ir/_graph_containers.py

```diff
@@ -21,18 +20,23 @@
     from onnxscript.ir import _core
 
 
-logger = logging.getLogger(__name__)
-
-
 class _GraphIO(collections.UserList["_core.Value"]):
     """The inputs and outputs of a Graph."""
 
     def __init__(self, graph: _core.Graph, initlist=None):
-        super().__init__(initlist)
         self._graph = graph
+        # Use a ref counter to track the number of references to each value
+        # in the input/output list. This is used to determine when to unset the graph
+        # reference in the value.
+        # Even though a duplicated value is invalid in inputs and not recommended in outputs,
+        # it is still possible to have duplicated inputs/outputs in an ONNX graph so we
+        # need to properly handle this case and maintain the graph reference properly.
+        self._ref_counter: collections.Counter[_core.Value] = collections.Counter()
```

Comment:
nit: It might be more efficient to simply recheck if a removed input is still in the input list (given than removing an input is an unlikely operation ... I doubt if there is much functionality today that actually does this). On the other hand, it may not matter much (adding inputs does happen, but probably not that much).